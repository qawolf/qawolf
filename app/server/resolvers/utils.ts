/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Transaction } from "knex";

import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findGroup } from "../models/group";
import { findSuite } from "../models/suite";
import { findTest } from "../models/test";
import { Team, Test, User } from "../types";

type EnsureGroupAccess = {
  group_id: string;
  logger: Logger;
  teams: Team[] | null;
  trx?: Transaction;
};

type EnsureSuiteAccess = {
  logger: Logger;
  suite_id: string;
  teams: Team[] | null;
};

type EnsureTeams = {
  logger: Logger;
  teams: Team[] | null;
};

type EnsureTeamAccess = {
  logger: Logger;
  team_id: string;
  teams: Team[] | null;
};

type EnsureTestAccess = {
  logger: Logger;
  teams: Team[] | null;
  test?: Test;
  test_id?: string;
};

type EnsureUser = {
  logger: Logger;
  user: User | null;
};

/**
 * @summary Throws if `teams` is falsy
 * @returns `teams`
 */
export const ensureTeams = ({ logger, teams }: EnsureTeams): Team[] => {
  const log = logger.prefix("ensureTeams");

  if (!teams) {
    log.error("teams not provided");
    throw new AuthenticationError("no teams");
  }

  return teams;
};

export const ensureGroupAccess = async ({
  group_id,
  logger,
  teams,
  trx,
}: EnsureGroupAccess): Promise<Team> => {
  const log = logger.prefix("ensureGroupAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access group", group_id);

  const group = await findGroup(group_id, { logger, trx });
  const selectedTeam = teams!.find((team) => group.team_id === team.id);

  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access group", group_id);
    throw new AuthenticationError("cannot access group");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `suite.team_id`. Can also
 *   throw if `suite_id` doesn't match any suite.
 * @returns The team from `teams` that matches `suite.team_id`, if we didn't throw.
 */
export const ensureSuiteAccess = async ({
  logger,
  teams,
  suite_id,
}: EnsureSuiteAccess): Promise<void> => {
  const log = logger.prefix("ensureSuiteAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access suite", suite_id);

  const suite = await findSuite(suite_id, { logger });

  if (!teamIds.includes(suite.team_id)) {
    log.error("teams", teamIds, "cannot access suite", suite_id);
    throw new Error("Teams cannot access suite");
  }
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `team_id`
 * @returns The team from `teams` that matches `team_id`, if we didn't throw.
 */
export const ensureTeamAccess = ({
  logger,
  team_id,
  teams,
}: EnsureTeamAccess): Team => {
  const log = logger.prefix("ensureTeamAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  const selectedTeam = teams!.find((team) => team.id === team_id);
  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access team", team_id);
    throw new AuthenticationError("cannot access team");
  }

  log.debug("ensured teams", teamIds, "can access team", team_id);

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `test.team_id`. Can also
 *   throw if `test_id` doesn't match any test.
 * @returns The team from `teams` that matches `test.team_id`, if we didn't throw.
 */
export const ensureTestAccess = async ({
  logger,
  teams,
  test,
  test_id,
}: EnsureTestAccess): Promise<Team> => {
  const log = logger.prefix("ensureTestAccess");

  const finalTeams = ensureTeams({ logger, teams });
  const teamIds = finalTeams.map((t) => t.id);

  log.debug("ensure teams", teamIds, "can access test", test?.id || test_id);

  const finalTest = test || (await findTest(test_id, { logger }));
  const team = finalTeams.find((t) => finalTest.team_id === t.id);

  if (!team) {
    log.error("teams", teamIds, "cannot access test", test?.id || test_id);
    throw new AuthenticationError("cannot access test");
  }

  return team;
};

/**
 * @summary Throws if `user` is falsy
 * @returns `user` if we didn't throw
 */
export const ensureUser = ({ logger, user }: EnsureUser): User => {
  const log = logger.prefix("ensureUser");

  if (!user) {
    log.error("user not provided");
    throw new AuthenticationError("no user");
  }

  return user;
};
