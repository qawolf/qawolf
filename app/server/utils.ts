import { randomBytes, scrypt } from "crypto";
import buildCuid from "cuid";
import * as EmailValidator from "email-validator";
import { v4 as uuidv4 } from "uuid";

import { ClientError } from "./errors";
import { Logger } from "./Logger";

type IsCorrectCode = {
  code: string;
  digest: string;
};

export type RunnerUrl = {
  id: string;
  location: string;
};

export const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
export const API_KEY_PREFIX = "qawolf_";

export const randomChoice = <T>(items: T[] | string): T | string => {
  const randomIndex = Math.floor(Math.random() * items.length);

  return items[randomIndex];
};

export const buildApiKey = (): string => {
  // remove dashes
  const uuid = uuidv4().split("-").join("");

  return `${API_KEY_PREFIX}${uuid}`;
};

// https://dev.to/farnabaz/hash-your-passwords-with-scrypt-using-nodejs-crypto-module-316k
export const buildDigest = async (code: string): Promise<string> => {
  return new Promise((resolve, reject) => {
    const salt = randomBytes(16).toString("hex");

    scrypt(code, salt, 64, (error, derivedKey) => {
      if (error) reject(error);
      resolve(salt + ":" + derivedKey.toString("hex"));
    });
  });
};

export const buildLoginCode = (length = 6): string => {
  let code = "";

  for (let i = 0; i < length; i++) {
    code += randomChoice(ALPHABET);
  }

  return code;
};

export const buildRunnerHost = (options: RunnerUrl): string => {
  if (options.location === "local") return `localhost`;

  return `${options.location}.qawolf.com/runner/${options.id}`;
};

export const buildRunnerStatusUrl = (options: RunnerUrl): string => {
  const protocol = options.location === "local" ? "http" : "https";
  return `${protocol}://${buildRunnerHost(options)}/.qawolf/runner/status`;
};

export const buildRunnerWsUrl = (options: RunnerUrl): string => {
  const protocol = options.location === "local" ? "ws" : "wss";
  return `${protocol}://${buildRunnerHost(options)}/.qawolf`;
};

export const cuid = (): string => buildCuid();

export const isCorrectCode = async ({
  code,
  digest,
}: IsCorrectCode): Promise<boolean> => {
  return new Promise((resolve, reject) => {
    const [salt, key] = digest.split(":");

    scrypt(code, salt, 64, (error, derivedKey) => {
      if (error) reject(error);
      resolve(key === derivedKey.toString("hex"));
    });
  });
};

// https://stackoverflow.com/a/1214753
export const minutesFromNow = (minutes = 0): string => {
  return new Date(Date.now() + minutes * 60000).toISOString();
};

export const validateEmail = (email: string, logger: Logger): void => {
  const isValid = EmailValidator.validate(email);
  if (isValid) return;

  logger.error(`email ${email} invalid`);
  throw new ClientError("invalid email address");
};
