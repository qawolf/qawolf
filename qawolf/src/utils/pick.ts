// Since we only need this for now, we can avoid downloading lodash
const pick = (
  obj: Record<any, any>,
  props: readonly string[]
): Record<any, any> => {
  const newObj = {};
  for (const prop of props) {
    if ({}.hasOwnProperty.call(obj, prop)) {
      newObj[prop] = obj[prop];
    }
  }
  return newObj;
};

export default pick;
