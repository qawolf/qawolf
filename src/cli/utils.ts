import { parse, UrlWithStringQuery } from 'url';

export const omitArgs = (args: string[], argsToOmit: string[]): string[] => {
  return args.filter(arg => !argsToOmit.some(skip => arg.startsWith(skip)));
};

export const parseUrl = (urlString: string): UrlWithStringQuery => {
  let url = parse(urlString);

  // prefix w/ https if a protocol is not provided
  if (!url.protocol) {
    url = parse(`https://${urlString}`);
  }

  if (!url.hostname) {
    throw new Error(`Invalid url ${urlString}`);
  }

  return url;
};
