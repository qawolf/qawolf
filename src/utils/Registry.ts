import { EventEmitter } from 'events';
import { Browser, BrowserContext } from 'playwright';
import * as qawolf from '../qawolf';

type RegistryData = {
  browser?: Browser;
  context?: BrowserContext;
  qawolf?: typeof qawolf;
  [key: string]: any;
};

export class Registry extends EventEmitter {
  private static _instance = new Registry();

  public static instance(): Registry {
    return this._instance;
  }

  private _data: RegistryData = {};

  public data(): RegistryData {
    return this._data;
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  _setValue(key: string, value: any): void {
    this._data[key] = value;
    this.emit('change');
  }

  public setBrowser(browser: Browser): void {
    this._setValue('browser', browser);
  }

  public setContext(context: BrowserContext): void {
    this._setValue('context', context);
  }

  public setQawolf(value: typeof qawolf): void {
    this._setValue('qawolf', value);
  }
}
