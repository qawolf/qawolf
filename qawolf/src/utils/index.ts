import pick from "./pick";

export { pick };
