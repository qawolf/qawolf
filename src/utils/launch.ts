import Debug from 'debug';
import { platform } from 'os';
import playwrightCore, {
  Browser,
  BrowserType,
  LaunchOptions as PlaywrightLaunchOptions,
} from 'playwright';
import { isNullOrUndefined } from 'util';
import { Registry } from './Registry';

const debug = Debug('qawolf:launch');

type BrowserName = 'chromium' | 'firefox' | 'webkit';

export type LaunchOptions = PlaywrightLaunchOptions & {
  browserName?: BrowserName;
};

const parseBool = (value: string | undefined): boolean => {
  const lowerCaseValue = (value || '').toLowerCase();
  return ['1', 't', 'true'].includes(lowerCaseValue);
};

export const parseBrowserName = (name?: string): BrowserName => {
  if (name === 'firefox' || name === 'webkit') return name;

  return 'chromium';
};

export const getBrowserType = (
  browserName: BrowserName,
): BrowserType<Browser> => {
  // We must use the browser type from the installed `playwright` or `playwright-browser` package,
  // and not `playwright` since they store different browser binaries.
  // See https://github.com/microsoft/playwright/issues/1191 for more details.
  let playwright: typeof playwrightCore;

  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    playwright = require('playwright');
  } catch (error) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      playwright = require(`playwright-${browserName}`);
    } catch (error) {
      throw new Error('qawolf requires playwright to be installed');
    }
  }

  return playwright[browserName];
};

export const getLaunchOptions = (
  options: LaunchOptions = {},
): LaunchOptions & { browserName: BrowserName } => {
  const launchOptions = { ...options };

  const headlessEnv = process.env.QAW_HEADLESS;
  if (isNullOrUndefined(options.headless) && !isNullOrUndefined(headlessEnv)) {
    launchOptions.headless = parseBool(headlessEnv);
  }

  const browserName = parseBrowserName(
    options.browserName || process.env.QAW_BROWSER,
  );

  const defaultArgs: string[] = [];

  if (browserName === 'chromium' && platform() === 'linux') {
    // We use --no-sandbox because we cannot change the USER for certain CIs (like GitHub).
    // "Ensure your Dockerfile does not set the USER instruction, otherwise you will not be able to access GITHUB_WORKSPACE"
    defaultArgs.push('--no-sandbox');
  }

  return {
    args: defaultArgs,
    // override args if they are provided
    ...launchOptions,
    browserName,
  };
};

export const launch = async (options: LaunchOptions = {}): Promise<Browser> => {
  const launchOptions = getLaunchOptions(options);
  debug('launch %j', launchOptions);

  const browser = await getBrowserType(launchOptions.browserName).launch(
    launchOptions,
  );

  Registry.instance().setBrowser(browser);

  return browser;
};
