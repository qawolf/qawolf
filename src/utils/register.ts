import { BrowserContext } from 'playwright';
import { basename, join } from 'path';
import { indexPages } from './context/indexPages';
import { saveArtifacts } from './context/saveArtifacts';
import { Registry } from './Registry';

export const getArtifactPath = (): string | null => {
  let artifactPath = process.env.QAW_ARTIFACT_PATH;
  if (!artifactPath) return null;

  if (require.main) {
    // store artifacts under the name of the main module, if there is one
    // ex. /artifacts/search.test.js
    artifactPath = join(artifactPath, basename(require.main.filename));
  }

  // store artifacts under the name of the browser being tested
  const browserName = process.env.QAW_BROWSER;
  if (browserName && artifactPath) {
    artifactPath = join(artifactPath, browserName);
  }

  return artifactPath;
};

export const register = async (context: BrowserContext): Promise<void> => {
  Registry.instance().setContext(context);

  const promises: Promise<any>[] = [];

  promises.push(indexPages(context));

  const artifactPath = getArtifactPath();
  if (artifactPath) {
    promises.push(saveArtifacts(context, artifactPath));
  }

  await Promise.all(promises);
};
