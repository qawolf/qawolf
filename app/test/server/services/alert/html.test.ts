import {
  buildInviteHtml,
  buildLoginCodeHtml,
  buildSuiteHtml,
} from "../../../../server/services/alert/html";
import { buildGroup, buildInvite, buildUser } from "../../utils";

const group = buildGroup({});

describe("buildInviteHtml", () => {
  it("builds html for invite", () => {
    const html = buildInviteHtml({
      ...buildInvite({}),
      creator_email: "spirit@qawolf.com",
      team_name: "My Team",
    });

    expect(html).toMatchSnapshot();
  });
});

describe("buildLoginCodeHtml", () => {
  it("builds html for login code", () => {
    const html = buildLoginCodeHtml({
      login_code: "SPIRIT",
      user: buildUser({}),
    });

    expect(html).toMatchSnapshot();
  });
});

describe("buildSuiteHtml", () => {
  it("builds html for failure email", () => {
    const html = buildSuiteHtml({
      group,
      runs: [
        {
          gif_url: "https://gif.gif",
          id: "runId",
          is_test_deleted: false,
          status: "pass",
          test_id: "testId",
          test_name: "logIn",
        },
        {
          gif_url: "https://gif.gif",
          id: "run2Id",
          is_test_deleted: false,
          status: "fail",
          test_id: "test2Id",
          test_name: "signUp",
        },
        {
          gif_url: null,
          id: "run3Id",
          is_test_deleted: false,
          status: "fail",
          test_id: "test3Id",
          test_name: "logOut",
        },
      ],
      suite_id: "suiteId",
    });

    expect(html).toMatchSnapshot();
  });

  it("builds html for success email", () => {
    const html = buildSuiteHtml({
      group,
      runs: [
        {
          gif_url: "https://gif.gif",
          id: "runId",
          is_test_deleted: false,
          status: "pass",
          test_id: "testId",
          test_name: "logIn",
        },
      ],
      suite_id: "suiteId",
    });

    expect(html).toMatchSnapshot();
  });
});
