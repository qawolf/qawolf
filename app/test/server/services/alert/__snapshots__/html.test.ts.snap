// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildInviteHtml builds html for invite 1`] = `
"<div style='color: #1C2F46; font-family: Helvetica, Arial, sans-serif;'>
  <table>
    <tr>
      <td><img src='https://storage.googleapis.com/docs.qawolf.com/website/logo_small.png' style='object-fit: contain; width: 48px;' /></td>
      <td><h1 style='display: inline-block; font-size: 24px; font-weight: normal; margin: 0 8px;'>QA Wolf</h1></td>
    </tr>
  </table>
  <div style='font-size: 16px; margin-top: 32px; text-align: center; width: 100%;'>
      <a style='color: inherit !important; font-weight: bold; text-decoration: none !important;'>spirit@qawolf.com</a>
      <p style='margin: 8px 0'>has invited you to join the team</p>
      <b>My Team</b>
      <br />
      <a href='http://localhost:3000/invite/inviteId' style='background: #44E5E7; border-radius: 8px; color: #1C2F46; display: inline-block; margin-top: 32px; font-weight: bold; padding: 16px 32px; text-decoration: none;'>Accept Invite</a>
  </div>
  <div style='margin-top: 48px;'>
    <img src='https://storage.googleapis.com/spirit-imgs/wolf-husky.png' style='display: block; margin: auto;' width=\\"80px\\" />
  </div>
  <div style=\\"border-top: 1px solid #979797; font-size: 12px; margin: 0 32px 48px; text-align: center;\\">
    <p style='color: #979797; margin: 48px 32px;'>QA Wolf is an end-to-end testing platform that helps teams ship confidently. Create, run, and share tests - all without leaving the browser.</p>
  </div>
  </div>"
`;

exports[`buildLoginCodeHtml builds html for login code 1`] = `
"<div style='color: #1C2F46; font-family: Helvetica, Arial, sans-serif;'>
  <table>
    <tr>
      <td><img src='https://storage.googleapis.com/docs.qawolf.com/website/logo_small.png' style='object-fit: contain; width: 48px;' /></td>
      <td><h1 style='display: inline-block; font-size: 24px; font-weight: normal; margin: 0 8px;'>QA Wolf</h1></td>
    </tr>
  </table>
  <div style='font-size: 16px; margin-top: 32px; text-align: center; width: 100%;'>
    <p>Enter the code below to log in to QA Wolf:</p>
    <p style=\\"background: #d2e0e2; border-radius: 8px; display: inline-block; font-size: 48px; font-weight: bold; margin: 8px 0; padding: 8px 16px;\\">SPI-RIT</p>
    <p>If you didnâ€™t request this email, you can safely ignore it.</p>
  </div>
  <div style='margin-top: 48px;'>
    <img src='https://storage.googleapis.com/spirit-imgs/wolf-brown.png' style='display: block; margin: auto;' width=\\"80px\\" />
  </div>
  <div style=\\"border-top: 1px solid #979797; font-size: 12px; margin: 0 32px 48px; text-align: center;\\">
    <p style='color: #979797; margin: 48px 32px;'>QA Wolf is an end-to-end testing platform that helps teams ship confidently. Create, run, and share tests - all without leaving the browser.</p>
  </div>
  </div>"
`;

exports[`buildSuiteHtml builds html for failure email 1`] = `"<p>Oh no! The following <a href='http://localhost:3000/tests/suiteId'>group1 tests</a> failed:<p><br /><a href='http://localhost:3000/run/run2Id'>signUp</a><br /><img src=\\"https://gif.gif\\" width=\\"320px\\" /><br /><a href='http://localhost:3000/run/run3Id'>logOut</a><br />"`;

exports[`buildSuiteHtml builds html for success email 1`] = `"<p>All good! Your <a href='http://localhost:3000/tests/suiteId'>group1 tests</a> all passed."`;
